PR for review

tilsong opened the pull request:
- 리뷰를 위한 PR입니다.

tilsong committed:
[add Url Repository, edit Url Service]

- Repository 인터페이스 및 구현체 InmemoryUrlRepository 생성
- 변환 유틸인 Base62Converter 추가됨
- 각 트랜잭션의 범위 설정(미완료)

tilsong committed:
add 230205

tilsong committed:
[edit UrlController parameter, path, InmemoryUrlRepository]

- UrlController의 api 경로 변경 dto를 pathvariable로 변경, EncodedUrlRequest dto 제거
- InmemoryUrlRepository의 에러 로직 변경

tilsong committed:
[edit UrlController path, add validation]

- 각 api의 path를 rest에 맞게 변경했습니다.
- api의 parameter에 대한 validation을 추가했습니다.

tilsong committed:
[edit UrlController api - Bad Request 조건 추가]

f-lab-fi commented on src/main/java/com/toy/shorturl/domain/ViewUrl.java:
```diff
@@ -0,0 +1,3 @@
+package com.toy.shorturl.domain;
+
+public record ViewUrl(int id) {}
```
ViewUrl을 Url에 포함하지 않고, 이렇게 따로 만들어주신 이유가 있을까요?

tilsong replied:
Url은 기존 Url과 단축 Url을 담고 있는 데이터의 형태이고, 따라서 각 Url의 조회 수를 기록하는 역할을 하는 데이터는 따로 존재해야 한다고 생각했습니다. 그래서 분리했는데, 결과적으로 코드를 더 불필요하게 복잡하게 만드는 원인이 되었네요ㅠ

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/Url/InmemoryUrlRepository.java:
```diff
@@ -0,0 +1,88 @@
+		String defaultUrl = url.getUrl();
+
+		synchronized (insertLock) {
+			// url 중복이면 throw
+			boolean isPresent = urlList.stream()
+				.filter(e -> e.getUrl().equals(defaultUrl))
+				.findFirst() // 있으면 에러 방출
+				.isPresent();
+			if (isPresent) {
+				throw new RuntimeException("이미 존재하는 URL입니다.");
+			}
```
findFirst()의 반환 타입은 옵셔널이죠?
이 옵셔널을 통해 아래 있는 if 문을 없앨 수 있을 것 같네요. :)

f-lab-fi commented on src/main/java/com/toy/shorturl/service/UrlService.java:
```diff
@@ -1,13 +1,94 @@
+
+		// save transaction
+		try {
+			Url newUrl = new Url(url);
+			int index = urlRepository.save(newUrl);
+
+			encodedUrl = Base62Converter.intTobase62(index);
+
+			urlRepository.update(index, encodedUrl);
```
인코딩할 때 index가 사용되고 있는데요, 이렇게 index(id 값)을 활용하는 이유가 있을까요?

tilsong replied:
인코딩하여 단축 Url을 만들 수 있는 값은 ArrayList에 데이터를 save한 뒤 얻게 되는 index 값 혹은 단축되지 않은 기본 Url이라고 생각했습니다.
그런데 Url의 경우, 여러 문자(특수 기호 등)가 섞일 수 있기 때문에 모든 문자를 고려하여 인코딩하기는 어려울 것 같다고 생각했습니다. 
반면 index 값은 양의 정수의 값이기 때문에, 규칙적인 패턴을 적용할 수 있을 것이라고 생각했습니다.

f-lab-fi replied:
넵 좋습니다.
랜덤하게 일정 길이 생성하는 것도 가능한데 논리만 있다면 괜찮습니다.

f-lab-fi commented on src/main/java/com/toy/shorturl/controller/UrlController.java:
```diff
@@ -1,29 +1,87 @@
+		log.info("create Url. request url: " + urlRequest.url());
+
+		String encodedUrl = urlService.addUrl(urlRequest.url());
+		if (encodedUrl != null) {
+			UrlResponse dto = new UrlResponse(encodedUrl);
+			return new ResponseEntity<>(dto, HttpStatus.CREATED);
 		}
-		return "잘못된 url입니다.";
+
+		return new ResponseEntity(HttpStatus.BAD_REQUEST);
```
encodedUrl가 null이 되는 경우는 어떤 경우에 발생하는건가요?

f-lab-fi replied:
https://github.com/tilsong/short-url/pull/1/files#diff-bfb2e13ce125d710af25dbf8545474362666bd61a2957546f02552c1636a73f3R68

여기서 예외가 발생해서 service에서 catch 후 null을 리턴한다고 이야기 하실 것 같은데, 예외를 좀 더 효과적으로 처리할 수 있는 방법은 없을까요?

f-lab-fi commented on src/main/java/com/toy/shorturl/controller/UrlController.java:
```diff
@@ -1,29 +1,87 @@
+	) throws URISyntaxException {
+		log.info("get Url. request encodedUrl: " + encodedUrl);
+
+		String decodedUrl = urlService.findUrlByEncodedUrl(encodedUrl);
+		if (decodedUrl != null) {
+			HttpHeaders httpHeaders = new HttpHeaders();
+			httpHeaders.setLocation(new URI(decodedUrl));
+
+			return new ResponseEntity(httpHeaders, HttpStatus.MOVED_PERMANENTLY);
+		}
+
+		return new ResponseEntity(HttpStatus.BAD_REQUEST);
```
위랑 같은 맥락, decodedUrl는 어떤 경우에 null 인거죠?

f-lab-fi commented on src/main/java/com/toy/shorturl/controller/UrlController.java:
```diff
@@ -1,29 +1,87 @@
+	@GetMapping("/viewcount/{encodedUrl}") // www.localhost:8080/viewcount/{encodedUrl}
+	public ResponseEntity<ViewcountResponse> getViewCountByEncodedUrl(
+		@PathVariable("encodedUrl") @NotBlank(message = "입력 값이 존재하지 않습니다.") String encodedUrl
+	) {
+		log.info("get ViewCountByEncodedUrl. request encodedUrl: " + encodedUrl);
+
+		int viewCount = urlService.getViewCount(encodedUrl);
+		if(viewCount != -1) {
+			ViewcountResponse dto = new ViewcountResponse(encodedUrl, viewCount);
+			return new ResponseEntity(dto, HttpStatus.OK);
+		};
+
+		return new ResponseEntity(HttpStatus.BAD_REQUEST);
+	}
```
viewCount가 -1인 경우는 어떤 경우 일까요?

(제가 똑같은걸 계속 물어보네요? ㅎㅎ)

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/ViewUrl/InmemoryViewUrlRepository.java:
```diff
@@ -0,0 +1,30 @@
+	public int countById(int id) {
+		return (int) viewcountList.stream()
+			.filter(e -> e.id() == id)
+			.count();
```
?? 이거 count 0이나 1만 나오는거 아닌가용

f-lab-fi commented on src/test/java/com/toy/shorturl/Domain/ViewUrl/ViewUrlRepositoryTest.java:
```diff
@@ -0,0 +1,32 @@
+	@AfterEach
+	void afterEach() { viewUrlRepository.clearStore();}
+
+	@Test
+	void saveAndCountById() {
+		int insertCount = 10000;
+		int expectCounting = 100;
+
+		for (int i = 0; i < insertCount; i++) {
+			ViewUrl viewUrl = new ViewUrl(i % 100); // id = 0~99, 각각 100개씩 save
+			viewUrlRepository.save(viewUrl);
+		}
+
+		int findCounting = viewUrlRepository.countById(10); // 0~99 중 하나의 id
+
+		assertThat(findCounting).isEqualTo(expectCounting);
+	}
```
뭔가 요구사항이 잘못 전달된 것 같은데,

> Optional) 단축된 URL에 대한 요청 횟수 저장 및 요청 횟수 조회 기능

이 말은
단축된 URL에 대한 요청 횟수가 저장되어야 한다는 의미지, 
단축된 URL에 대한 저장 횟수가 저장되어야 한다는 의미가 아닙니다.

f-lab-fi replied:
지금 구현하신 코드랑 테스트 코드를 보니, 저장 횟수를 저장하고 계시는 것 같아서 이야기 드립니다.
제가 잘못 이해한거면 알려주세요.

tilsong replied:
맨 위에서 답변한 것처럼 각 Url의 조회 수를 기록하는 역할을 하는 데이터는 따로 존재하고 따로 관리할 수 있어야 한다고 생각해서 저렇게 작성했었던 것 같습니다.
만들 때는 ViewUrl의 필드에  사용자 정보도 추가하면 확장에 용이하겠지? 하고 생각해서 만들었는데, 지금 생각하면 필요하지 않은 고려사항이었던 것 같습니다..ㅠ

f-lab-fi commented on src/main/java/com/toy/shorturl/service/UrlService.java:
```diff
@@ -1,13 +1,94 @@
+	}
+
+	public int getViewCount(String encodedUrl) {
+		log.info("get viewCount.");
+
+		int index = Base62Converter.base62ToInt(encodedUrl);
+
+		// index validation
+		try {
+			Url oneByIndex = urlRepository.findOneByIndex(index);
+		} catch (RuntimeException re) {
+			log.error(re.getMessage());
+			return -1;
+		}
+
+		int count = viewUrlRepository.countById(index);
 
-		return url;
+		return count;
 	}
```
이 코드가 지금 코드의 문제를 제대로 보여주고 있습니다.
Url와 ViewUrl을 나눔으로서 불필요하게 2번 조회를 만들고 있고, 코드가 불필요하게 복잡해지고 있죠?

tilsong replied:
넵ㅠㅠ

tilsong committed:
[Remove ViewUrl, 예외 처리 적용]

- 불필요하게 적용되었던 ViewUrl이 제거되었습니다.
  - 그에 따른 레포지토리 및 테스트 코드도 제거되었습니다.
  - Url에 viewCount 필드를 추가하고, 이를 활용하게 되었습니다.

- 특정 상황에 맞는 예외가 추가되었습니다.
  - Url 추가 시 중복이면 발생하는 DuplicateUrlException
  - Url 검색 시 없으면 발생하는 NoSuchUrlException

- 추가된 예외를 활용하여 @RestControllerAdvice를 통한 전역 예외 핸들러가 추가되었습니다.
  - 이를 통해 서비스에서 컨트롤러로 주던 에러 상황에서의 특정 값이 사라졌습니다. 단지 컨트롤러로 추가된 예외를 던지게 됩니다.

f-lab-fi commented on src/main/java/com/toy/shorturl/service/UrlService.java:
```diff
@@ -6,47 +6,37 @@
 
-		String encodedUrl = null;
+		Url newUrl;
 
-		// save transaction
 		try {
-			Url newUrl = new Url(url);
-			int index = urlRepository.save(newUrl);
-
-			encodedUrl = Base62Converter.intTobase62(index);
+			var savingUrl = new Url(url);
+			int index = urlRepository.save(savingUrl);
 
+			var encodedUrl = Base62Converter.intTobase62(index);
 			urlRepository.update(index, encodedUrl);
-		} catch (RuntimeException re) {
-			log.error(re.getMessage());
-		}
 
-		// find transaction
-		Url newUrl;
-		try {
 			newUrl = urlRepository.findOneByUrl(url);
-		} catch (RuntimeException re) {
-			log.error(re.getMessage());
-			return null;
+		} catch (DuplicateUrlException | NoSuchUrlException e) {
+			log.error(e.getMessage());
+			throw e;
```
이건 제가 지난번에 미처 못봤던 포인트인데, 여기서는 예외 발생시 로그 레벨을 error로 찍어주시고 [아래에서는](https://github.com/tilsong/short-url/pull/1/files/b9eb3c0142ab2c9307f8f0f469d2a33125b35058..da0362a0a7e5cd3b380fd5f91d83efef0c7f2fb7#diff-38ac5737fef8ce180e82f50a0c4f5506b0040ca84d36f17b53f6a3ed4fc0ad39R56-R59) info로 찍어주셨는데요.
은석님이 생각하시는 어떤 기준이 있는걸까요?

tilsong replied:
예외가 발생한 것이기 때문에 모두 error 레벨로 출력하는 것이 맞을 것 같습니다. info로 찍힌 부분 수정하겠습니다.

f-lab-fi replied:
아뇨아뇨 단순히 그렇게 생각할 문제가 아닙니다!
로그 레벨에 대해선 멘토링 시간에 다시 이야기 한번 해봐요.
제가 깜빡할 수도 있으니 이야기 해야한다고 말해주세요.

tilsong replied:
넵 알겠습니다~!

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/Url/UrlRepository.java:
```diff
@@ -8,6 +8,8 @@ public interface UrlRepository {
 
     void update(int index, String encodedUrl);
 
+    void UpdateViewCount(int index);
+
```
왜 대문자로 시작하죠? ^^

tilsong replied:
아앗...!

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/Url/InmemoryUrlRepository.java:
```diff
@@ -48,25 +50,31 @@ public void update(int index, String encodedUrl) {
 	}
 
-	 public Url findOneByIndex(int index) throws RuntimeException {
+	public void UpdateViewCount(int index) {
+		var url = findOneByIndex(index);
+		url.increaseViewCount();
+
+		log.info("Url viewCount updated. " + "id: " + index + " " + url);
+	}
```
단축 URL이 한번 사용될 때마다 뷰 카운트를 올리는 행위는 다분히 비즈니스 로직에 해당할만한 행위입니다.
이 행위를 레포지토리에 넣는게 적합한걸까요?
만약 InmemoryUrlRepository 말고 또 다른 구현체가 생긴다면 이 코드가 계속 중복되어야 하지 않을까요?

tilsong committed:
[Edit UrlService, add ViewCountService, 그 외]

- 예외 로그 info, error를 구분해서 작성했습니다.
  - 서비스 단에서 catch 절 안에 들어가는 예외가 발생할 경우, 이를 나타내는 로그는 모두 log.error를 사용하도록 했습니다.
  - 반면 메소드가 정상적으로 수행되면서 기록되는 로그는 log.info를 사용하도록 했습니다.

- UpdateViewCount 오타가 수정되었습니다. (첫 글자  대 -> 소문자)

- viewCount를 업데이트하는 로직을 서비스로 이동시켰습니다.
  - repository는 단순히 index와 새로운 viewCount만을 받아 이를 update하는 로직을 수행합니다.
  - 서비스 단에서 update 방식에 대한 내용을 다루게 됩니다. 상세 내용은 아래에 있습니다.

- 특정 url을 조회할 경우 수행되는 조회수 업데이트에 대한 로직을 비동기 처리했습니다.
  - 클라이어트의 url 조회에 대한 반환값은 동기적으로 반환됩니다. 반면 조회수 업데이트 로직은 해당 요청과 직접적으로 상관 있지 않으므로 비동기적으로 작동되도록 했습니다.
  - 비동기 처리를 위한 설정 파일을 추가했습니다. 현재 애플리케이션이 작동되는 cpu 자원에 따라 스레드 개수가 변경될 수 있도록 했습니다.
  - 비동기 메서드 처리를 위한 새로운 서비스를 추가했습니다.
  - 조회 수 업데이트 시 동기화 이슈를 고려하여, synchronized를 적용했고, 락 스트라이핑 방식을 적용했습니다.

- Get /url/{encodedUrl} 및 Get /viewcount/{encodedUrl} 요청에 대한 validation을 추가했습니다.
  - 기존에는 Base62Converter 모듈의 양식에 맞지 않는 데이터(알파벳, 숫자가 아닐 경우)가 들어와 예외가 발생할 수 있었는데, 정규 표현식을 통한 validation을 통해 해결되었습니다.

- @RequiredArgsConstructor 을 사용했습니다.
  - 기존 생성자 주입을 위한 코드들을 사용하지 않게 되었습니다.

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/Url/InmemoryUrlRepository.java:
```diff
@@ -50,9 +50,9 @@ public void update(int index, String encodedUrl) {
+	public void updateViewCount(int index, long viewCount) {
 		var url = findOneByIndex(index);
-		url.increaseViewCount();
+		url.setViewCount(viewCount);
```
만약에 DB를 사용하는 Repository가 생겨도 지금 이 인터페이스(메서드 시그니처) 그대로 사용하실 것 같으신가요?

tilsong replied:
이 부분은 어떻게 접근해야 할지 잘 모르겠습니다..ㅠㅠ
단순히 Url의 viewcount 필드를 수정하는 sql을 만든다면 index와 viewcount를 사용하게 될 것 같아 위의 메서드 시그니처를 사용했습니다. 어떻게 접근해야 할까요?

tilsong replied:
만약 위와 같은 메서드 명을 쓰면 필드 갯수 만큼 메소드 갯수가 늘어나겠구나.. 라는 생각이 들었습니다. 그래서 index와 하나의 필드를 인자로 주어 메서드 하나당 하나의 필드를 수정하는 것이 아닌, 변경할 데이터와 기존 데이터를 모두 포함한 dto와 같은 객체를 통해 Url의 데이터를 일괄 수정하도록 하는 것이 더 낫겠다고 생각하게 되었습니다.

f-lab-fi replied:
말씀하신 것처럼 메서드가 계속 늘어나는 문제도 있고, 엔티티의 값을 변경시키는 행위를 레포지토리 구현체에서 하면 `url.setViewCount(viewCount);` 같은 코드는 계속 중복될겁니다. 이게 중복되는게 의미가 있을까요~?

tilsong replied:
엔티티의 값을 서비스 레이어에서 변경하고, 레포지토리 구현체에서는 단순히 받은 값을 저장하는 역할만 하면 좋을 것 같습니다! 레포지토리가 레포지토리의 역할만 하도록 해야할 것 같습니다.

f-lab-fi commented on src/main/java/com/toy/shorturl/service/UrlService.java:
```diff
@@ -12,15 +16,34 @@
+			log.error(e.getMessage());
+			log.error("id-" + index +"에 대한 추가 조회 수가 기록되지 않았습니다.");
+			// 로그만 기록하고 에러를 던지지 않음.
+		}
```
ㅋㅋ 이거 굳이 비동기적으로 처리하면서 락까지 거는 이유는 뭔가용?
코드 자체도 복잡해지고, 이렇게 따로 처리하는 오버헤드가 더 클 것 같은데..

tilsong replied:
만들 때 생각한 것은 

조회수 증가를 위해 Url의 viewcount 필드를 업데이트 해줄 때 기존에 저장된 Url의 viewcount 값을 가져오고, 여기에 1을 더한 값을 repository에 업데이트하는 과정을 거쳐야 한다고 생각습니다. 그래서 두 스레드가 동시에 같은 viewcount 값을 가져와 viewcount를 업데이트한다면 동기화 문제가 발생할 수 있으므로 동기화 처리가 필요하다고 생각되어.. synchronized를 사용하게 되었습니다.

비동기로 구현한 것은 조회수 업데이트를 위한 동기화 때문에 단축 url에 대한 클라이언트의 요청이 늦어져서는 안된다고 생각했기 때문입니다..ㅠ

혹시 이 부분도 ViewUrl 처럼 너무 생각을 앞서 나간 것일까요?ㅠㅠ

f-lab-fi replied:
아뇨 너무 앞서나간거라기보다, 이게 이 애플리케이션의 UX를 좌우하는 중요한 부분이었다면 성능을 신경 쓰는게 맞습니다.
그런데 한번 생각해보죠.

어떤 요청을 리다이렉트 시키는 과정에서 viewCount를 업데이트 시키는 행위는 비용이 많이 드는 작업일까요?
상황에 따라 비용이 많이 들 수도 있겠지만, 적어도 현재 메모리상에서 데이터를 관리하고 있을 때는 비용이 아주 작을겁니다.
오히려 비용을 가장 많이 차지하는 부분은 클라이언트와 서버 사이의 요청이 오고 가는데 사용되는 네트워크 시간 일 거에요.
따라서 이 과정은 이렇게 복잡하게 만들기에는 현재 상태에선 오버 엔지니어링이 될 가능성이 높다는 겁니다.
이건 저장소를 데이터베이스로 바꿨어도 마찬가지에요. 은석님이 생각한 것처럼 그렇게 큰 비용이 들지 않습니다.
설령 큰 비용이 든다는걸 확인했고, 이렇게 개선하면 해결 할 수 있다는걸 확인하고 나서 개선하면 됩니다.

지금은 우선 코드를 간단하고, 가독성 있게 짜는게 집중하세요.

tilsong replied:
넵 이해했습니다. 말씀해주신 대로 코드 다시 작성해보겠습니다!

f-lab-fi commented on src/main/java/com/toy/shorturl/service/ViewCountService.java:
```diff
@@ -0,0 +1,41 @@
+		Arrays.fill(locks, new Object());
+	}
+
+	@Async("threadPoolTaskExecutor")
+	public void addViewCount(int index) {
+		try {
+			synchronized (locks[index % locks.length]) {
+				Url url = urlRepository.findOneByIndex(index);
+				long currentViewCount = url.getViewCount();
+				urlRepository.updateViewCount(index, ++currentViewCount);
+			}
+		} catch (NoSuchUrlException e) {
+			log.error(e.getMessage());
+			log.error("id-" + index +"에 대한 추가 조회 수가 기록되지 않았습니다.");
+			// 로그만 기록하고 에러를 던지지 않음.
+		}
+	}
```
이 코드는 왜 UrlService에도 있고 여기도 있나용?

tilsong replied:
여기 있는 게 맞고 UrlService에 있는 거는 제가 테스트하고 깜빡한 것 같습니다!

tilsong committed:
[Edit log level, InmemoryUrlRepository, remove async logic]

- 에러 레벨 수정
  - 기존에 log.error 기록했던 예외 상황들이 시스템에 영향을 줄만한 예외들로 판단되지는 않기 때문에, 이를 log.info로 변경했습니다.

- InmemoryUrlRepository의 update 관련 메소드 변경
  - 기존에는 필드 마다 별도의 update 메소드를 만들어주었습니다. 이를 update 메소드 하나로 대체했습니다.
  - 기존에는 repository 내에서 엔티티를 변경했으나, repository 이전 레이어(여기서는 service layer)에서 엔티티를 미리 변경하고, repository는 변경된 엔티티를 받아 기존 데이터에 update 시키는 역할만 하게 되었습니다.
  - 서비스 단에서 Url 객체를 생성할 때, 빌더 패턴을 사용하도록 했습니다. 이를 통해 repository에 변경된 객체 자체를 전달할 수 있게 되었습니다.

- 비동기 처리 관련 내용 제거
  - 기존에 분리했던 viewcount 업데이트 관련 비동기 처리 로직을 제거했습니다.

f-lab-fi commented on src/main/java/com/toy/shorturl/repository/Url/UrlRepository.java:
```diff
@@ -6,13 +6,11 @@ public interface UrlRepository {
     int save(Url url);
 
-    void update(int index, String encodedUrl);
+    void update(int index, Url url);
 
-    void updateViewCount(int index, long currentViewCount);
+    Url findOneByEncodedUrl(String encodedUrl);
 
     Url findOneByIndex(int index);
```
Repository에 index가 있는게 자연스러운걸까요?
지금 쓰고 있는 List 기반의 Repository 말고 다른 종류(Map이나 Database 등)의 구현체가 추가된다고 해도 지금의 인터페이스가 적합할지 고민해보시기 바랍니다.
